File: Baekjoon/gold/1918.py
import sys

input = lambda: sys.stdin.readline().rstrip()

stack = []

expression = input()

priority = {"(": 3, "*": 2, "/": 2, "+": 1, "-": 1}

for char in expression:
    if "A" <= char <= "Z":
        print(char, end="")

    elif len(stack) == 0:
        stack.append(char)

    elif char == ")":
        while len(stack) > 0 and stack[-1] != "(":
            print(stack.pop(), end="")
        stack.pop()

    elif priority[char] > priority[stack[-1]]:
        stack.append(char)

    elif priority[char] <= priority[stack[-1]]:
        while len(stack) > 0 and priority[char] <= priority[stack[-1]] and stack[-1] != "(":
            print(stack.pop(), end="")
        stack.append(char)

while stack:
    print(stack.pop(), end="")

File: Baekjoon/gold/1918.cpp
#include <bits/stdc++.h>
using namespace std;


int main() {
    string expression;
    cin >> expression;

    map<char, int> priority = { {'(', 3}, {'*', 2}, {'/', 2}, {'+', 1}, {'-', 1} };

    stack<char> s;

    for (int i = 0; i < expression.size(); i++) {
        if ('A' <= expression[i] and expression[i] <= 'Z') {
            cout << expression[i];
            continue;
        }

        if (expression[i] == ')') {
            while (s.top() != '(') {
                cout << s.top();
                s.pop();
            }
            s.pop();
            continue;
        }

        while (not s.empty() and priority[s.top()] >= priority[expression[i]] and s.top() != '(') {
            cout << s.top();
            s.pop();
        }

        s.push(expression[i]);
    }

    while (not s.empty()) {
        cout << s.top();
        s.pop();
    }

    return 0;
}